portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java

```diff
@@ -0,0 +1,50 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+@WebServlet("/like")
+public class LikeServlet extends HttpServlet {
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {    
+    String likeComment = request.getParameter("like");
+    if (likeComment==null || likeComment.isEmpty()){
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Comment must be a non-empty string.");
+      return;
+    }
+
+    Filter eqContent = new FilterPredicate("content", FilterOperator.EQUAL, likeComment);
+    Query query = new Query("Comment").setFilter(eqContent);
+    PreparedQuery pq = this.datastore.prepare(query);
+    Entity commentEntity;
+    try{
+      commentEntity = pq.asSingleEntity();
+    } catch(PreparedQuery.TooManyResultsException e){
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Comment is not unique.");
+      return;
+    }
+    if (commentEntity==null){
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Comment not found in the database");
+      return;
+    }
+
+    // Update property of the existing comment entity
+    commentEntity.setProperty("timestamp", (long)System.currentTimeMillis());
+    commentEntity.setProperty("like", (long)commentEntity.getProperty("like")+(long)1);
```

Comment:
What happens if two people have your website open and like the same comment at the exact same time? ;)

Reply:
Hopefully that doesn't happen. I am counting on Datastore to figure out this concurrency issue with [Transactions](https://cloud.google.com/appengine/docs/standard/java/datastore/api-overview#Transactions). It seems Datastore can handle max 1 transaction per second on the same entity. So if two people do like the same comment at the exact same time, I assume there would be some delay for either one to see the final result, but the final result is at least accurate.

Reply:
Every datastore write occurs in a transaction, and those are serialized. But that's not enough in our case, because it says nothing about parallel reads. Imagine the following scenario.

1. User A likes comment X
2. (at the same time) User B likes comment X
3. Servlet (in thread 1) receives User A's request and looks up number of likes. X has 1 like
4. Servlet (in thread 2) receives User B's request and looks up number of likes. X has one like
5. Servlet (in thread 1) updates the comment entity with likes=1+1
6. Servlet (in thread 2) waits for transaction started by thread 1 to complete, then updates the comment entity with likes=1+1
7. Final like count: 2 (should be 3)

You can test this yourself ( and I suggest you do ) by inserting a call to Thread.sleep in your handler between the datastore read and write, then acting as User A and B in two separate tabs.

We can fix this issue with transactions though. Right now your code uses them implicitly. But we can use a transaction explicitly and wrap the get and put calls in the same transaction. This way thread 2 will not be able to read the current number of likes until thread 1 has finished updating them.

There is an example in the Datastore documentation: https://cloud.google.com/appengine/docs/standard/java/datastore/transactions

Reply:
Thanks for the detailed example! It really helps me understand why the end result is off and how bundling read and write can prevent that. Will implement it. 

On the same page you provided, it is also suggested "counter-sharding" as a solution to this issue. Seems like an overkill for this application though. 

And lastly, if you have any good resources for a beginner to learn more about threads please share!

Reply:
Definitely agree that sharding is overkill in this situation. If your website goes viral we should talk about how to improve performance, though.

Reply:
I wonder when that will be 🤔 
Transaction is added.